opentracing_instrumentation/client_hooks/requests.py

```diff
@@ -25,44 +25,53 @@
 import logging
 import urllib.parse
 
+from opentracing.ext import tags
 from ..request_context import get_current_span
 from ..http_client import AbstractRequestWrapper
 from ..http_client import before_http_request
 from ..http_client import split_host_and_port
-from ._singleton import singleton
+from ._patcher import Patcher
 
 log = logging.getLogger(__name__)
 
 # Try to save the original entry points
 try:
-    import requests.sessions
     import requests.adapters
 except ImportError:  # pragma: no cover
     pass
 else:
     _HTTPAdapter_send = requests.adapters.HTTPAdapter.send
 
 
-@singleton
-def install_patches():
-    try:
-        import requests.sessions
-        import requests.adapters
-    except ImportError:  # pragma: no cover
-        return
-
-    def send_wrapper(self, request, **kwargs):
-        """Wraps HTTPAdapter.send"""
-
-        request_wrapper = RequestWrapper(request=request)
-        span = before_http_request(
-            request=request_wrapper,
-            current_span_extractor=get_current_span)
-        with span:
-            resp = _HTTPAdapter_send(self, request, **kwargs)
-            if hasattr(resp, 'status_code') and resp.status_code is not None:
-                span.set_tag('http.status_code', resp.status_code)
-        return resp
+class RequestsPatcher(Patcher):
+
+    applicable = '_HTTPAdapter_send' in globals()
+    response_handler_hook = None
```

Comment:
given the design of this class, shouldn't this also be an instance variable instead of class field?

Reply:
No, it's not required.

And also it's currently convenient to override `response_handler_hook` with some handler if you want to implement some custom patcher, e.g.
```python
class CustomRequestsPatcher(RequestsPatcher):

    def response_handler_hook(self, response, span):
        pass  # some handling here
```
or
```python
class CustomRequestsPatcher(RequestsPatcher):

    @staticmethod
    def response_handler_hook(response, span):
        pass  # some handling here
```

On the other hand, in case when you have something like
```python
class RequestsPatcher(Patcher):

    def __init__():
        self.response_handler_hook = None
```
it makes you to use `set_response_handler_hook` method when you already have your `CustomRequestsPatcher` or to override `__init__` which is not so convenient.

Just to be sure that we are on the same page:
`set_response_handler_hook` doesn't modify `response_handler_hook` attribute of the class. Instead, it sets attribute for an instance, so it doesn't affect the class.
```python
>>> import mock
>>> from opentracing_instrumentation.client_hooks.requests import RequestsPatcher
>>> foo_patcher = RequestsPatcher()
>>> bar_patcher = RequestsPatcher()
>>> foo_patcher.set_response_handler_hook(mock.Mock(name='foo'))
>>> bar_patcher.set_response_handler_hook(mock.Mock(name='bar'))
>>> str(RequestsPatcher.response_handler_hook)
'None'
>>> foo_patcher.response_handler_hook
<Mock name='foo' id='140467355335424'>
>>> bar_patcher.response_handler_hook
<Mock name='bar' id='140467610463424'>
```
The patcher uses attribute of the class as a default value only.

Reply:
+1